src/Optimizer/Transformer/AutoExtensions.php

```diff
@@ -0,0 +1,584 @@
+     */
+    private function addRequiredExtensionByAttributes(Document $document, Element $node, $extensionScripts)
+    {
+        if (! $node->hasAttributes()) {
+            return $extensionScripts;
+        }
+
+        // Look for forms.
+        if ($node->tagName === Tag::FORM) {
+            $extensionScripts = $this->maybeAddExtension($document, $extensionScripts, Extension::FORM);
+        }
+
+        // true if we need to import amp-bind.
+        $usesAmpBind = false;
+
+        // Associative array of new attribute name with data-amp-bind- prefix and attribute value pair.
+        $newAttributes = [];
+
+        // Populate all attributes for the current node tag name.
+        $nodeAttributeList = $this->getTagAttributeList($node->tagName);
+
+        $globalAttributes = $this->spec->attributeLists()->get(GlobalAttrs::ID);
+
+        foreach ($node->attributes as $attribute) {
+            // Add attribute dependencies (e.g. amp-img => lightbox => amp-lightbox-gallery).
+            if (array_key_exists($attribute->name, self::MANUAL_ATTRIBUTE_TO_EXTENSION_MAPPING)) {
+                $extensionScripts = $this->maybeAddExtension(
+                    $document,
+                    $extensionScripts,
+                    self::MANUAL_ATTRIBUTE_TO_EXTENSION_MAPPING[$attribute->name]
+                );
+            }
+
+            // Element attributes that require amp extensions (e.g. amp-video[dock]).
+            if (! empty($nodeAttributeList[$attribute->name][SpecRule::REQUIRES_EXTENSION])) {
+                $requiresExtensions = $nodeAttributeList[$attribute->name][SpecRule::REQUIRES_EXTENSION];
+                foreach ($requiresExtensions as $requiresExtension) {
+                    $extensionScripts = $this->maybeAddExtension($document, $extensionScripts, $requiresExtension);
+                }
+            }
+
+            // Attribute that requires AMP components (e.g. amp-fx).
+            if ($globalAttributes->has($attribute->name)) {
+                $attr = $globalAttributes->get($attribute->name);
+
+                if (isset($attr[SpecRule::REQUIRES_EXTENSION])) {
+                    foreach ($attr[SpecRule::REQUIRES_EXTENSION] as $requiresExtension) {
+                        $extensionScripts = $this->maybeAddExtension($document, $extensionScripts, $requiresExtension);
+                    }
+                }
+            }
+
+            // Check for amp-bind attribute bindings.
+            if (strpos($attribute->name, '[') === 0 || strpos($attribute->name, Amp::BIND_DATA_ATTR_PREFIX) === 0) {
+                $usesAmpBind = true;
+            }
+
+            /**
+             * EXPERIMENTAL FEATURE: Rewrite short-form `bindtext` to `data-amp-bind-text` to avoid false-positives we
+             * check for each tag only the supported bindable attributes (e.g. for a div only bindtext, but not
+             * bindvalue).
+             */
```

Comment:
Rephrasing this a bit to read better. Also note that this is not a docblock, as it doesn't document a specific element. Multi-line comments start with `/*`, not with `/**`.
```suggestion
            /*
             * EXPERIMENTAL FEATURE: Rewrite short-form `bindtext` to `data-amp-bind-text`.
             *
             * To avoid false-positives, we only check the supported bindable attributes for each tag (e.g. for a div
             * only bindtext, but not bindvalue).
             */
```